br/pkg/restore/data.go

```diff
@@ -292,50 +300,82 @@ func (recovery *Recovery) WaitApply(ctx context.Context) (err error) {
+	for _, v := range recovery.StoreMetas {
+		for _, m := range v.RegionMetas {
+			// insert the keyRanges
+			if _, ok := regions[m.RegionId]; !ok {
+				regions[m.RegionId] = struct{}{}
+				keyRanges = append(keyRanges, tikvstore.KeyRange{
+					StartKey: m.StartKey,
+					EndKey:   m.EndKey,
+				})
+			}
+		}
+	}
+
+	return keyRanges
+}
+
+// prepare the region for flashback the data, the purpose is to stop region service, put region in flashback state
+func (recovery *Recovery) PrepareFlashbackToVersion(ctx context.Context, keyRanges []tikvstore.KeyRange) (err error) {
 	eg, ectx := errgroup.WithContext(ctx)
-	totalStores := len(recovery.allStores)
-	workers := utils.NewWorkerPool(uint(mathutil.Min(totalStores, common.MaxStoreConcurrency)), "resolve data from tikv")
+	// since we do not know the how many region we will recover, use a ratio regions/per tikv as progress
+	incRatio := len(keyRanges) / len(recovery.allStores)
+	workers := utils.NewWorkerPool(uint(recovery.concurrency), "prepare the region")
 
-	// TODO: what if the resolved data take long time take long time?, it look we need some handling here, at least some retry may necessary
-	// TODO: what if the network disturbing, a retry machanism may need here
-	for _, store := range recovery.allStores {
+	for i, r := range keyRanges {
 		if err := ectx.Err(); err != nil {
 			break
 		}
-		storeAddr := getStoreAddress(recovery.allStores, store.Id)
-		storeId := store.Id
 		workers.ApplyOnErrorGroup(eg, func() error {
-			recoveryClient, conn, err := recovery.newRecoveryClient(ectx, storeAddr)
+			_, err := ddl.SendPrepareFlashbackToVersionRPC(ctx, recovery.mgr.GetStorage().(tikv.Storage), r)
 			if err != nil {
 				return errors.Trace(err)
 			}
-			defer conn.Close()
-			log.Info("resolve data to tikv", zap.String("tikv address", storeAddr), zap.Uint64("store id", storeId))
-			req := &recovpb.ResolveKvDataRequest{ResolvedTs: resolvedTs}
-			stream, err := recoveryClient.ResolveKvData(ectx, req)
+			log.Debug("prepare region done", zap.ByteString("region_start_key", r.StartKey), zap.ByteString("region_end_key", r.EndKey))
+			return nil
+		})
+
+		if i%incRatio == 0 {
+			recovery.progress.Inc()
+		}
+	}
+	log.Info("prepare region flashback complete")
+	// Wait for all region to prepare flashback
+	return eg.Wait()
+}
+
+// flashback the region data to resolvedTs
+func (recovery *Recovery) FlashbackToVersion(ctx context.Context, keyRanges []tikvstore.KeyRange, resolvedTs uint64, commitTs uint64) (err error) {
+	eg, ectx := errgroup.WithContext(ctx)
+	incRatio := len(keyRanges) / len(recovery.allStores)
+	workers := utils.NewWorkerPool(uint(recovery.concurrency), "flashback the region")
+
+	for i, r := range keyRanges {
+		if err := ectx.Err(); err != nil {
+			break
+		}
+		workers.ApplyOnErrorGroup(eg, func() error {
+			_, err := ddl.SendFlashbackToVersionRPC(ctx, recovery.mgr.GetStorage().(tikv.Storage), resolvedTs, commitTs-1, commitTs, r)
```

Comment:
If br paniced, what will happenned with commitTs? Use the new one or the old one? If the commitTS is different for regions, it's acceptable or not?

Reply:
1. User can not use BR to do EBS snapshot backup and restore. the solution only available in TiDB-Operator
2. the first restore (restore_ebs_meta) will check PD ts vs resolved_ts, ensure the correct one is used.
3. in this version, BR panic or other ERROR, the BR job will fail. TiDB-Operator can launch BR Job again. restore volume or TiKV pod will not be used. The volume will be regenerated from the snapshot.
so that commitTS is always acceptable from my  view